timelock-corruption-detection/src/main/java/com/palantir/history/PaxosLogHistoryProvider.java

```diff
@@ -61,16 +80,131 @@ private Long getOrInsertVerificationState(NamespaceAndUseCase namespaceAndUseCas
                 namespaceAndUseCase.namespace(), namespaceAndUseCase.useCase());
     }
 
-    public List<CompletePaxosHistoryForNamespaceAndUseCase> getHistory() {
+    /**
+     * Gets history from all the nodes for all unique namespace, useCase tuples since last verified sequence numbers
+     *
+     * @throws  RuntimeException
+     *          if fails to fetch history from all remote servers
+     */
+    public List<CompletePaxosHistoryForNamespaceAndUseCase> getHistory() throws RuntimeException {
+
+        /* used to load history since last verified sequence numbers for all unique namespace and use case tuples */
         Map<NamespaceAndUseCase, Long> lastVerifiedSequences = KeyedStream
                 .of(getNamespaceAndUseCaseTuples().stream())
                 .map(namespaceAndUseCase -> verificationProgressStateCache.computeIfAbsent(namespaceAndUseCase,
                         this::getOrInsertVerificationState))
                 .collectToMap();
+
+        /**
+         * The history queries are built from the lastVerifiedSequences map above,
+         * required for conjure endpoints to load remote history
+         */
```

Comment:
_Generally_, and I think this also does apply here: when functions reach this kind of complexity, you should consider extracting some of the individual calls to methods with descriptive names. (In this way you usually won't need comments as well.)